server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java

```diff
@@ -0,0 +1,402 @@
+              checkAuthorization(principal, expression, resourceKeys);
+            } catch (IOException e) {
+              // This is probably because we read partial data.
+              LOGGER.warn("Error parsing payload: {}", e.getMessage());
+            }
+          }
+        } else {
+          LOGGER.warn("Skipping content-type: {}", req.contentType());
+        }
+      });
+
+      return delegate.serve(ctx, peekReq);
+    }
+  }
+
+  private static Object findPayloadValue(String key, Map<String, Object> payload) {
+    // TODO: investigate better object traversal functionality
+    String[] args = key.split("[.]", 2);
+    if (args.length == 1) {
+      return payload.get(args[0]);
+    } else {
+      if (payload.get(args[0]) instanceof Map) {
+        @SuppressWarnings("unchecked")
+        Map<String, Object> value = (Map<String, Object>) payload.get(args[0]);
+        return findPayloadValue(args[1], value);
+      } else {
+        return null;
+      }
+    }
+  }
+
+  private void checkAuthorization(UUID principal, String expression, Map<ResourceType, Object> resourceKeys) {
+    LOGGER.debug("resourceKeys = {}", resourceKeys);
+
+    Map<ResourceType, Object> resourceIds = mapResourceKeys(resourceKeys);
+
+    if (!resourceIds.keySet().containsAll(resourceKeys.keySet())) {
+      LOGGER.warn("Some resource keys have unresolved ids.");
+    }
+
+    LOGGER.debug("resourceIds = {}", resourceIds);
+
+    if (!evaluator.evaluate(principal, expression, resourceIds)) {
+      throw new BaseException(ErrorCode.PERMISSION_DENIED, "Access denied.");
+    }
+  }
+
+  private Map<ResourceType, Object> mapResourceKeys(Map<ResourceType, Object> resourceKeys) {
+    Map<ResourceType, Object> resourceIds = new HashMap<>();
+
+    if (resourceKeys.containsKey(CATALOG) && resourceKeys.containsKey(SCHEMA) && resourceKeys.containsKey(TABLE)) {
+      String fullName = resourceKeys.get(CATALOG) + "." + resourceKeys.get(SCHEMA) + "." + resourceKeys.get(TABLE);
+      TableInfo table = TableRepository.getInstance().getTable(fullName);
+      resourceIds.put(TABLE, UUID.fromString(table.getTableId()));
+    }
+
+    // If only TABLE is specified, assuming its value is a full table name (including catalog and schema)
+    if (!resourceKeys.containsKey(CATALOG) && !resourceKeys.containsKey(SCHEMA) && resourceKeys.containsKey(TABLE)) {
+      String fullName = (String) resourceKeys.get(TABLE);
+      TableInfo table = TableRepository.getInstance().getTable(fullName);
+      String fullSchemaName = table.getCatalogName() + "." + table.getSchemaName();
+      SchemaInfo schema = SchemaRepository.getInstance().getSchema(fullSchemaName);
+      CatalogInfo catalog = CatalogRepository.getInstance().getCatalog(table.getCatalogName());
+      resourceIds.put(TABLE, UUID.fromString(table.getTableId()));
+      resourceIds.put(SCHEMA, UUID.fromString(schema.getSchemaId()));
+      resourceIds.put(CATALOG, UUID.fromString(catalog.getId()));
```

Comment:
why do we take resouceIds for each level of the namespace when full name of table is provided vs when each level is provided?

Reply:
In the other case, when multiple keys are configured, those get set separately later in the method.